Fix logging levels

gavinleeedwards opened the pull request:
# Motivation and Context
There are a lot of errors/exceptions being reported in Splunk that aren't actual errors (e.g. someone typing in an enrolment code wrong isn't a problem with the application).  This work changes the logging levels to 'info' for exceptions that have no negative impact.

**This work only covers the logging levels - there's another piece of work to improve and  standardise the way that exceptions are raised and errors are logged across all services but a card has been raised to cover it. **

# What has changed
- Made some error messages more explicit to help when diagnosing errors in Splunk.
- Changed log levels to 'info' where appropriate
- Set up a new ClientError exception to log some exceptions at 'info' level
- Added new tests to cover the ClientError exception

# How to test?
- Tests should pass
- Generate a few exceptions.  They should be logging at 'info' level where appropriate.

Commit:
Fix logging levels

shahi645 commented on ras_party/controllers/account_controller.py:
```diff
@@ -65,11 +65,11 @@ def post_respondent(party, tran, session):
     iac = request_iac(party['enrolmentCode'])
     if not iac.get('active'):
-        raise RasError("Enrolment code is not active.", status=400)
+        raise ClientError("Enrolment code is not active.", status=400)
```
👍

ajmaddaford replied:
The error messages passed to `ClientError` aren't used - they don't appear in the logs.

shahi645 commented on ras_party/controllers/account_controller.py:
```diff
@@ -230,13 +233,14 @@ def change_respondent_password(token, payload, tran, session):
     respondent = query_respondent_by_email(email_address, session)
     if not respondent:
-        raise RasError("Respondent does not exist.", status=404)
+        raise RasError("Respondent does not exist. Unable to change respondent's password",
```
@gavinleeedwards Do we want info level here?

shahi645 replied:
ignore, just seen the `ClientError`

ajmaddaford replied:
I think we need to be careful about trying to be too helpful in the log messages and repeating information already in the log message. For example here, the context of the respondent not existing should be clear if we're logging the url, without needing to add "Unable to change respondent's password"

gavinleeedwards replied:
I don't necessarily agree that being too helpful is a bad thing.  The issue here is that `Respondent does not exist` could come from several places in the code so making sort of messages less generic makes it quicker to pinpoint where any issues might be occurring.

ajmaddaford replied:
I think we'll end up with a lot of duplication. 'Expired email verification token' and 'Unknown email verification token' appear multiple times but haven't had any more logging added. If we applied the same 'Unable to change respondent's password' to all of those log lines there would be a lot of repetition. If the url was being logged we'd see 'Respondent does not exist' along with `url=/respondents/change_password/<token>, making it clear what action was being performed when the error occurred.

shahi645 commented on ras_party/error_handlers.py:
```diff
@@ -5,13 +5,21 @@
 
 
+@blueprint.app_errorhandler(ClientError)
+def client_error(error):
```
👍

andrewmil commented on ras_party/controllers/account_controller.py:
```diff
@@ -297,7 +301,7 @@ def request_password_change(payload, session):
         except RasNotifyError:
             # Note: intentionally suppresses exception
-            logger.error('Error sending notification email for party_id', party_id=party_id)
+            logger.error('Error sending reset password email for party_id', party_id=party_id)
```
For consistency/clarity might be better to use `respondent_id=party_id`

gemmai95 commented on ras_party/controllers/account_controller.py:
```diff
@@ -297,7 +301,7 @@ def request_password_change(payload, session):
                 email_address, personalisation, str(party_id))
         except RasNotifyError:
             # Note: intentionally suppresses exception
-            logger.error('Error sending notification email for party_id', party_id=party_id)
```
Does this party_id need changed to respondent_id?

gavinleeedwards replied:
Changed the log message to respondent_id to make it clearer 👍

ajmaddaford replied:
I think this log message is a little misleading - we don't send emails from the party service. We should be explicit in our log messages, something like "Error sending request to notify gateway" would be clearer here.

andrewmil reviewed (commented):
Looks good left a couple of very small comments but it's might not be for this PR :+1:
Just running against acceptance tests now then I'll approve

Commit:
Amend log message to make it more explicit

ajmaddaford commented on ras_party/controllers/account_controller.py:
```diff
@@ -65,11 +65,11 @@ def post_respondent(party, tran, session):
     existing = query_respondent_by_email(party['emailAddress'], session)
     if existing:
-        raise RasError("Email address already exists.", status=400, party_uuid=party.get('party_uuid', None))
+        raise ClientError("Email address already exists.", status=400, party_uuid=party.get('party_uuid', None))
```
Although `ClientError` persists the kwargs in a property, these aren't used when the exception is caught and the log line printed i.e. extra arguments like `party_uuid` won't be output in the logs.

ajmaddaford commented on ras_party/controllers/account_controller.py:
```diff
@@ -208,13 +211,13 @@ def verify_token(token, session):
     respondent = query_respondent_by_email(email_address, session)
     if not respondent:
-        raise RasError("Respondent does not exist.", status=404)
+        raise RasError("Respondent does not exist.  Unable to verify token.", status=404)
```
Why don't all instances of 4XX status codes use ClientError?

gavinleeedwards replied:
It depends on the circumstances.  For example, all `Respondent does not exist` errors are 404s.  In some cases, it doesn't matter if the exception is raised so it's a ClientError (e.g. if a user mistypes a respondemt id).  In other cases, it's an issue if the respondent isn't found if we're expecting it to, so RasError.

ajmaddaford replied:
I think by doing this we're choosing which exception to raise based on what log level we want the message output at. This would be better dealt with by the service calling this endpoint to interpret the result and how to log it.

Also, by raising it as a `RasError` it's logged as an exception with the text 'Uncaught exception'.

Commit:
Amend all 404 errors to ClientError, pass kwargs to exception logs and log url in exceptions

gavinleeedwards commented:
@ajmaddaford I've changed all 400s to ClientError, added the url to log messages, changed the RasNotify error message and made sure that the kwargs are now passed through to the exception logs. 👍

Commit:
Fix failing tests and linting errors

ajmaddaford commented on ras_party/support/session_decorator.py:
```diff
@@ -21,12 +21,16 @@ def handle_session(f, args, kwargs):
+        raise
     except RasError:
-        logger.error(f"Rolling back database session due to failure executing function")
+        logger.info(f"Rolling back database session due to failure executing function")
```
Minor, but this doesn't need to be an f-string

gavinleeedwards replied:
Done 👍

ajmaddaford commented on ras_party/exceptions.py:
```diff
@@ -1,3 +1,15 @@
+
+    def __init__(self, errors, status=None, **kwargs):
+        self.errors = errors if isinstance(errors, list) else [errors]
+        self.status_code = status or RasPartyError.status_code
```
This will set the default error code to 500 (`RasPartyError` is a `RasError`). If we defaulted status to 400 rather than `None` we could drop the `status_code = 400`:
```python
class ClientError(Exception):

    def __init__(self, errors, status=400, **kwargs):
        self.errors = errors if isinstance(errors, list) else [errors]
        self.status_code = status
        self.kwargs = kwargs
```

gavinleeedwards replied:
Done 👍

ajmaddaford commented on ras_party/controllers/account_controller.py:
```diff
@@ -155,7 +158,8 @@ def change_respondent_enrolment_status(payload, session):
     respondent = query_respondent_by_party_uuid(respondent_id, session)
     if not respondent:
-        raise RasError("Respondent does not exist.", status=404)
+        raise ClientError("Respondent does not exist.  Unable to change enrolment status",
```
I think we can drop the "Unable to change enrolment status" now that we're logging the url

gavinleeedwards replied:
Done 👍

ajmaddaford commented on ras_party/controllers/account_controller.py:
```diff
@@ -311,7 +316,8 @@ def change_respondent_account_status(payload, party_id, session):
     respondent = query_respondent_by_party_uuid(party_id, session)
     if not respondent:
-        raise RasError("Unable to find respondent account", status=404)
+        raise ClientError("Respondent does not exist.  Unable to change account status.", respondent_id=party_id,
```
I think we can drop the "Unable to change account status" now that we're logging the url

gavinleeedwards replied:
Done 👍

ajmaddaford commented on ras_party/controllers/account_controller.py:
```diff
@@ -58,18 +58,20 @@ def post_respondent(party, tran, session):
     existing = query_respondent_by_email(party['emailAddress'], session)
     if existing:
-        raise RasError("Email address already exists.", status=400, party_uuid=party.get('party_uuid', None))
+        raise ClientError("Email address already exists.",
```
The existing log lines are not consistent on whether they end with a full stop or not. I say they shouldn't as the message should be a single event. For any we're amending, I think we should drop the trailing full stop.

gavinleeedwards replied:
Done - removed all full stops to keep log messages consistent.

ajmaddaford commented on ras_party/controllers/account_controller.py:
```diff
@@ -257,7 +262,7 @@ def change_respondent_password(token, payload, tran, session):
+        logger.error('Error sending notification email', respondent_id=party_id)
 
     # This ensures the log message is only written once the DB transaction is committed
     tran.on_success(lambda: logger.info('Respondent has changed their password', party_id=party_id))
```
This log line is still `party_id`, should it be updated to `respondent_id` as above?

gavinleeedwards replied:
Done 👍

ajmaddaford commented on ras_party/controllers/account_controller.py:
```diff
@@ -297,7 +302,7 @@ def request_password_change(payload, session):
         except RasNotifyError:
             # Note: intentionally suppresses exception
-            logger.error('Error sending notification email for party_id', party_id=party_id)
+            logger.error('Error sending request to Notify Gateway for respondent_id', respondent_id=party_id)
```
The " for respondent_id" isn't needed in the log event, as it will be clear from the log line including `respondent_id`.

gavinleeedwards replied:
Done 👍

Commit:
PR changes - amend log messages and default ClientError to status 400

gavinleeedwards reviewed (commented):
@ajmaddaford All done and ready for re-review 👍

ajmaddaford commented on ras_party/controllers/account_controller.py:
```diff
@@ -269,13 +274,13 @@ def change_respondent_password(token, payload, tran, session):
-        raise RasError("Respondent does not exist.", status=404)
+        raise ClientError("Respondent does not exist", status=404)
 
     logger.debug("Requesting password change", party_id=respondent.party_uuid)
```
This appears in the log as:
```
{"party_id": "UUID('b73ad0e0-c83d-454b-93ac-12ed58c741e2')", "event": "Requesting password change", ... }
```
We should wrap `respondent.party_uuid` with `str` in all places where we're accessing the uuid from the database object.

ajmaddaford commented on ras_party/error_handlers.py:
```diff
@@ -2,21 +2,37 @@
+    logger.exception('Uncaught exception',
+                     url=request.url,
+                     errors=error.to_dict(),
+                     **error.kwargs,
```
Minor, but I think `**error.kwargs` should be the last argument, consistent with how it's used in function definitions.

Commit:
Move kwargs to last argument in exception logging and cast all uuids to strings for logs messages

ajmaddaford commented on ras_party/controllers/account_controller.py:
```diff
@@ -294,17 +294,18 @@ def request_password_change(payload, session):
-        party_id = respondent.party_uuid
         try:
             NotifyGateway(current_app.config).request_password_change(
                 email_address, personalisation, str(party_id))
```
`party_id` is now a string so we can drop the `str` here

ajmaddaford commented on ras_party/controllers/account_controller.py:
```diff
@@ -89,7 +89,7 @@ def post_respondent(party, tran, session):
     if not business:
         raise ClientError("Could not locate business when creating business association",
-                          business_id=business_id,
+                          business_id=str(business_id),
```
`business_id` came from payload, so should already be a string

ajmaddaford commented on ras_party/controllers/account_controller.py:
```diff
@@ -159,7 +159,7 @@ def change_respondent_enrolment_status(payload, session):
     if not respondent:
         raise ClientError("Respondent does not exist",
-                          respondent_id=respondent_id, status=404)
+                          respondent_id=str(respondent_id), status=404)
```
`respondent_id ` came from payload, so should already be a string

ajmaddaford commented on ras_party/controllers/account_controller.py:
```diff
@@ -316,7 +317,7 @@ def change_respondent_account_status(payload, party_id, session):
     respondent = query_respondent_by_party_uuid(party_id, session)
     if not respondent:
-        raise ClientError("Respondent does not exist", respondent_id=party_id,
+        raise ClientError("Respondent does not exist", respondent_id=str(party_id),
```
`respondent_id ` comes from the url, so should already be a string

ajmaddaford commented on ras_party/controllers/account_controller.py:
```diff
@@ -410,7 +411,7 @@ def resend_verification_email(party_uuid, session):
     :return: make_response
     """
-    logger.debug('Attempting to resend verification_email', party_uuid=party_uuid)
+    logger.debug('Attempting to resend verification_email', party_uuid=str(party_uuid))
```
`party_uuid ` comes from the url, so should already be a string

ajmaddaford commented on ras_party/controllers/account_controller.py:
```diff
@@ -462,7 +463,7 @@ def add_new_survey_for_respondent(payload, tran, session):
         if not business:
             raise ClientError("Could not locate business when creating business association",
-                              business_id=business_id,
+                              business_id=str(business_id),
```
`business_id` comes from the response from the case service, so should already be a string

ajmaddaford commented on ras_party/controllers/account_controller.py:
```diff
@@ -477,26 +478,26 @@ def add_new_survey_for_respondent(payload, tran, session):
     tran.on_success(lambda: logger.info('Respondent has enroled to survey for business',
                                         survey_name=survey_name,
-                                        business=business_id))
+                                        business=str(business_id)))
```
`business_id` comes from the response from the case service, so should already be a string

ajmaddaford commented on ras_party/controllers/business_controller.py:
```diff
@@ -67,7 +67,7 @@ def get_business_by_id(party_uuid, session, verbose=False, collection_exercise_i
     business = query_business_by_party_uuid(party_uuid, session)
     if not business:
-        raise ClientError("Business with party id does not exist", party_uuid=party_uuid, status=404)
+        raise ClientError("Business with party id does not exist", party_uuid=str(party_uuid), status=404)
```
`party_uuid` comes from the url, so should already be a string

ajmaddaford commented on ras_party/controllers/party_controller.py:
```diff
@@ -58,12 +58,12 @@ def get_party_by_id(sample_unit_type, id, session):
         business = query_business_by_party_uuid(id, session)
         if not business:
-            raise ClientError("Business with id does not exist", business_id=id, status=404)
+            raise ClientError("Business with id does not exist", business_id=str(id), status=404)
```
`id` comes from the url, so should already be a string

ajmaddaford commented on ras_party/controllers/party_controller.py:
```diff
@@ -58,12 +58,12 @@ def get_party_by_id(sample_unit_type, id, session):
         respondent = query_respondent_by_party_uuid(id, session)
         if not respondent:
-            raise ClientError("Respondent with id does not exist", respondent_id=id, status=404)
+            raise ClientError("Respondent with id does not exist", respondent_id=str(id), status=404)
```
`id` comes from the url, so should already be a string

ajmaddaford commented on ras_party/controllers/respondent_controller.py:
```diff
@@ -78,7 +78,7 @@ def change_respondent_details(respondent_data, respondent_id, session):
     respondent = query_respondent_by_party_uuid(respondent_id, session)
     if not respondent:
-        raise ClientError("Respondent id does not exist", respondent_id=respondent_id, status=404)
+        raise ClientError("Respondent id does not exist", respondent_id=str(respondent_id), status=404)
```
`respondent_id ` comes from the url, so should already be a string

gavinleeedwards replied:
@ajmaddaford  All done.

Commit:
Remove casts to string where they are not needed

Commit:
Fix codacy issue
